### ISSUE TITLE
Type caching issue because sql_identifier has no regtype

### ISSUE THREAD

**Issue by Cito** [PyGreSQL/PyGreSQL#65]
The following bug was reported by Justin Pryzby on the mailing list 2021-09-10.

**To reproduce**:

Run the following code against a PostgreSQL 12 database:
```python
from pg import DB

db = DB('postgres')
q = db.query("""
    select array_agg(column_name) as cols
    from information_schema.columns
""")
print(q.onescalar())
```

There are two problems here: First, the result is expected to be an array of strings, but it is a raw array string instead. Second, the query produces an error in the logs under the hood: `"_sql_identifier" does not exist`.

**Cause**:

In PostgreSQL 12, object-name columns in the information_schema views are now declared as being of domain type `sql_identifier` (they had been `varchar` before, which had been unproblematic). Since PyGreSQL does not know the `sql_identifier` type, it tries to create and cache it. To do this, it requests the basic information about the type from the database, which includes the internal type name and the registered type name. The registered type name is determined using the expression `typename::regtype` which throws an error for  `sql_identifier` because it has no registered type name. You can verify this as follows:
```sql
select 'sql_identifier'::regtype
```

**Solution/ToDo**:

As a solution, Justin proposed to use `to_regtype()` instead of `::regtype`. In fact, this function will return NULL rather than throwing an error if the name is not found. However, care must be taken as it only exists since PostgreSQL 9.4. Also, the DBType would be initialized with None and have regtype=None in this case. This should be documented and verified that this cannot cause adverse effects. Or maybe another string should be used instead of None. Also, it should be ensured that everything works with `DB.use_regtypes(False)` and `DB.use_regtypes(True)`. Since PostgreSQL versions before 9.5 have reached their EOL already, we probably don't need to fix this for PostgreSQL < 9.4, particularly since the issue is only crucial since PostgreSQL 12.

**Comment by Cito** [comment]
After some more comments from Justin on the mailing list the problem is now much clearer and we habe a better solution:

**Actual cause**:

`sql_identifier` is just an example. The issue is that pg tries to use `typname::regtype`, which fails when the typnamespace (schema) is not in the search path. For example, with default search path, these fail:
```sql
SELECT typname FROM pg_type WHERE typnamespace='information_schema'::regnamespace;
SELECT typname::regtype FROM pg_type WHERE oid='information_schema.views'::regtype;
-- ERROR:  type "views" does not exist
```
For instance 
```sql
SELECT '{1}'::information_schema.cardinal_number[]; -- "cardinal_number" does not exist
SET search_path=information_schema;
SELECT '{1}'::information_schema.cardinal_number[]; -- now it works
```
More examples:
```sql
-- this works
SELECT typname::regtype FROM pg_type WHERE typname!='any' AND NOT oid::regtype::text~'information_schema|any';
-- even for user types
CREATE TABLE t(); SELECT 't'::regtype;
-- but this gives an error
CREATE SCHEMA s; CREATE TABLE s.t2(); SELECT 't2'::regtype;
```

**Solution/ToDo:**

The solution is to use `oid::regtype` (rather than `typname::regtype`).

The [docs show](https://www.postgresql.org/docs/9.3/catalog-pg-type.html) that's available since v9.3. It actually works since 8.4 but wasn't documented until [Dec 2012](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=160701f6a935d5b0440751c3cb3f70bb59cf5eb1).

BTW, I think the casts like `::regtype` should be written as `::pg_catalog.regtype`, as they do it in Postgres itself, e.g. `describe.c` and `pg_dump.c`.

**Comment by Cito** [comment]
Fixed in febc87907eab0b453a635dfad131a37c8446e30d.
